Create new project

matthewbastien opened the pull request:
Implementation of the feature request #754.

The workflow is as follows:
1. User runs the command ```Swift: New Project```
2. A prompt appears to select the project type. Available project types are retrieved by parsing the output of ```swift package init --help```.
3. An open folder dialog appears to select the root folder for the new project.
4. A quick input appears to enter a name for the project.
5. The project folder is created on disk and ```swift package init``` is executed to create the project based on the given inputs.
6. A modal appears asking whether or not the user would like to open the newly created folder in the current window or a new one. An automatic action for this step can be configured with the ```swift.openAfterCreateProject``` setting. This is essentially the same modal used by the ```Git: Clone``` command.

I've also added a fairly basic test to make sure that the parsing of ```swift package init --help``` is functioning as intended.

Feedback is greatly appreciated! This is my first commit to this project and I definitely don't know all the ins and outs to the code base.

daveyc123 commented on src/commands.ts:
```diff
@@ -682,6 +834,7 @@ function updateAfterError(result: boolean, folderContext: FolderContext) {
  */
 export function register(ctx: WorkspaceContext) {
     ctx.subscriptions.push(
+        vscode.commands.registerCommand("swift.createProject", () => createProject(ctx)),
```
createProject() has a few instances where it returns undefined. What's the experience for the user in these cases? Should createProject be providing guidance/errors to the user in the form of notifications?

matthewbastien replied:
The places where createProject() returns undefined correspond to the user cancelling dialogs. I don't think we need any notifications in this case.

daveyc123 replied:
👍 What's the behaviour from vscode for exceptions (e.g. the execSwift or mkdir fail)

matthewbastien replied:
It pops up a modal with the error message from the exception.

adam-fowler replied:
Have you verified what the message is for say `mkdir` failing? Would be good to make sure it is human readable.

matthewbastien replied:
Good question! I tried it out myself and it looks like the error message is fairly self explanatory:

<img width="268" alt="Screenshot 2024-04-22 at 1 05 31 PM" src="https://github.com/swift-server/vscode-swift/assets/65295938/db2af8b2-43ab-4901-9aab-96d9b9ab632c">

adam-fowler commented on src/commands.ts:
```diff
@@ -91,6 +92,139 @@ export async function updateDependencies(ctx: WorkspaceContext) {
+        },
+    });
+    if (!projectName) {
+        return undefined;
```
Can we move this check inside the validateInput function of showInputBox

matthewbastien replied:
I've added some more validation here, but the check for undefined is still necessary for cancellation.

adam-fowler commented on src/commands.ts:
```diff
@@ -91,6 +92,139 @@ export async function updateDependencies(ctx: WorkspaceContext) {
+    // Prompt the user whether or not they want to open the newly created project
+    const isWorkspaceOpened = !!vscode.workspace.workspaceFolders;
+    const config = vscode.workspace.getConfiguration("swift");
+    const openAfterCreate = config.get<string>("openAfterCreateProject");
```
Can we add configuration settings to configuration.ts

matthewbastien replied:
Done.

adam-fowler commented on src/commands.ts:
```diff
@@ -91,6 +92,139 @@ export async function updateDependencies(ctx: WorkspaceContext) {
+
+    // Use swift package manager to initialize the swift project
+    await withDelayedProgress(
+        {
```
Given there is no progress reported here, any reason you can't use `WorkspaceContext.statusItem.showStatusWhileRunning()`

matthewbastien replied:
I'd expect that 99% of the time we would never even show the notification. In my testing it never showed up unless I reduced the timeout. However, if it does take longer than expected, I want it to be a prominent notification rather than something hidden away in the status bar. That way users know that something is actually happening in the background.

That being said, this does add extra code that will need to be maintained. I'm not too strongly attached to this opinion and can just use showStatusWhileRunning() instead if that makes more sense to you.

adam-fowler commented on src/toolchain/toolchain.ts:
```diff
@@ -178,6 +187,47 @@ export class SwiftToolchain {
+     */
+    public async getProjectTemplates(): Promise<SwiftProjectTemplate[]> {
+        const { stdout } = await execSwift(["package", "init", "--help"], "default");
+        const lines = stdout.split(/\r?\n/g);
```
This is some funky way to get project templates, but yeah can't see any other way of doing it.

I almost wonder if it is worthwhile rolling our own outside of SwiftPM. Would love to give users the ability to create their own templates. Maybe using Mustache templating to allow users to add in customisation. But that'd be for another task I think.

matthewbastien replied:
Yeah I agree. I'd honestly like an option in Swift Package Manager to get these as JSON rather than doing the crazy command line parsing. However, we still need to support Swift versions <6.

adam-fowler reviewed (commented):
In general looks pretty good, I've added a few comments

adam-fowler commented:
@swift-server-bot test this please

adam-fowler commented:
@swift-server-bot add to allowlist

adam-fowler commented:
CI is returning a bunch of errors for older versions of swift. New features can probably be gated on the last three versions of swift. Get it working for 5.8 if it works for all the earlier swift versions as well you are done. Otherwise Add a new contextKey `newProjectAvailable` in contextKeys.ts and use the contextKey to control availability of the command. At startup you can check the swift version and set it to false if the swift version is less than 5.8

award999 commented on package.json:
```diff
@@ -37,6 +37,11 @@
     "commands": [
+      {
+        "command": "swift.createProject",
+        "title": "New Project",
```
suggestion: "New" in title but "create" in ID, maybe make consistent?

suggestion: Typically have ellipsis suffix if the command will further prompt user for input

matthewbastien replied:
Done.

award999 commented on src/commands.ts:
```diff
@@ -91,6 +92,139 @@ export async function updateDependencies(ctx: WorkspaceContext) {
+    }
+
+    if (action === "open") {
+        vscode.commands.executeCommand("vscode.openFolder", projectUri, { forceReuseWindow: true });
```
suggestion: can we await or .catch to not leave a floating promise?

matthewbastien replied:
Good catch!

award999 commented on src/commands.ts:
```diff
@@ -91,6 +92,139 @@ export async function updateDependencies(ctx: WorkspaceContext) {
+        1000
+    );
+
+    // Prompt the user whether or not they want to open the newly created project
```
suggestion: maybe we can separate the creation from the opening of the project? The function is getting a little long

adam-fowler replied:
If we're going to do that maybe we should break it out of commands.swift and give it its own file

matthewbastien replied:
I'm inclined to leave it where it is for now. I'm not sure when we'd even re-use these parts. The project creation aspect is literally just a mkdir() + execSwift() call. If we do end up re-using some of the code down the road then we can split it out.

award999 reviewed (commented):
a few minor comments :)

matthewbastien commented on package.json:
```diff
@@ -160,6 +165,23 @@
       {
         "title": "Swift",
         "properties": {
+          "swift.openAfterCreateProject": {
```
Done. I put it at the end of the settings since I can't imagine this being a high runner use case.

matthewbastien commented on src/utilities/utilities.ts:
```diff
@@ -321,3 +321,78 @@ export function expandFilePathTilda(filepath: string): string {
+/**
+ * A wrapper around {@link vscode.Progress} that allows for delaying progress reporting.
+ */
+class ProgressWrapper<T> implements vscode.Progress<T> {
```
Done.

matthewbastien commented:
> CI is returning a bunch of errors for older versions of swift. New features can probably be gated on the last three versions of swift. Get it working for 5.8 if it works for all the earlier swift versions as well you are done. Otherwise Add a new contextKey `newProjectAvailable` in contextKeys.ts and use the contextKey to control availability of the command. At startup you can check the swift version and set it to false if the swift version is less than 5.8

The output of `swift package init --help` changes pretty drastically in versions <5.8.0. I've added the context key for enablement of the command, but we still need to support users being able to create a project in cases where the Swift extension is not activated yet (e.g. when no workspace is open). As such, I've also added a `swift.isActivated` context key to detect this. An error is shown to the user if they try to use the create project command with an unsupported swift version.

adam-fowler reviewed (approved):
I think we are good here
